### PR TITLE
[libraries/User] Add checks and error logging to prevent notices and warnings - Former #2964

### PR DESCRIPTION
**Pull request by xlecours**
Same as https://github.com/aces/Loris/pull/2964 but rebased

- [x] Ensure array values exist before passing them into SQL queries

- [x] Provide more verbose error logging to assist developers in debugging 

- [x] Prevent generation of vague PHP Notices when these values did not exist but were used 

**Note**: These issues largely came up while using `request_account` so it's possible that some of the execution flow in that code is responsible (e.g. looking for examiner associations before the user is actually created?). As such, the underlying issue may exist somewhere other than this class.

Still, these checks should exist and if the issues lie in other places in the code base, this will aid us in discovering and fixing them.

### CODE REVIEW

File: php/libraries/User.class.inc [hunk 1]
```diff
@@ -66,19 +70,33 @@ class User extends UserPermissions
...
-            "SELECT CenterID FROM user_psc_rel upr
-                        WHERE upr.UserID= :UID",
-            array('UID' => $row['ID'])
-        );
+        $user_cid = array();
+        if (!isset($row['ID'])) {
+            error_log(
+                "Could not get user center ID(s) because'
+                . ' user DB query value 'ID' is not set."
+            );
```
Comment (driusan) [inline php/libraries/User.class.inc]: I don't think this is possible. The users.ID column isn't nullable... which also means this should probably be an exception, not an error if it does need to stay for some reason.

File: php/libraries/User.class.inc [hunk 2]
```diff
@@ -66,19 +70,33 @@ class User extends UserPermissions
...
+        } else {
+            $user_centerID_query =  $DB->pselect(
+                "SELECT CenterID FROM user_psc_rel upr
+                WHERE upr.UserID= :UID",
+                array('UID' => $row['ID'])
+            );
```
Comment (driusan) [inline php/libraries/User.class.inc]: maybe this would be a good use for @ridz1208's new pselectCol function?

File: php/libraries/User.class.inc [hunk 3]
```diff
@@ -43,6 +43,10 @@ class User extends UserPermissions
      */
     static function &factory($username)
     {
+        if (is_null($username)) {
+            throw new LorisException('User factory called with null username');
+            return null;
```
Comment (driusan) [inline php/libraries/User.class.inc]: this return line is unreachable. And what if $username == ""?

File: php/libraries/User.class.inc (hunk 2 above)
Comment (xlecours) [reply php/libraries/User.class.inc]: https://github.com/aces/Loris/pull/3165 is not merged yet

**Comment by xlecours** [comment]
Can't create users with that "fix". I don't know it there was a bug in the first place. This will be easier to implement when the module is split in multiple pages doing one thing instead of a simple page for Create user, Edit user, My preferences and making breakfast.

**Pull request closed without merging**